File: image-smoother/shaun-gy.cpp
//with gy's idea, runtime 384ms, faster than 6%, mem less than 50%
class Solution {
public:
    vector<vector<int>> imageSmoother(vector<vector<int>>& M) {
        if (M.empty() || M[0].empty()) return {};
        int m = M.size(), n = M[0].size();
        vector<vector<int>> res(m,vector<int>(n));
        
        /*
        array faster
        int deltas [8][2] = {{-1,0}, {1,0},  //up and down
                     {0,-1},{0,1}, //left and right
                     {1,-1},{-1,1},  //diagonal /
                     {-1,-1},{1,1}}; //diagonal \
                     */
        
        vector<vector<int>> deltas = {{-1,0}, {1,0},  //up and down
                     {0,-1},{0,1}, //left and right
                     {1,-1},{-1,1},  //diagonal /
                     {-1,-1},{1,1}}; //diagonal \
        
        for (int i = 0; i < m; ++i) {
            for (int j = 0; j < n; ++j) {
                int count = 1, sum = M[i][j]; //include itself
                for (auto d: deltas) {
                    int x = i + d[0], y = j + d[1];
                    if (x < 0 || x >= m || y < 0 || y >= n) continue;
                    sum += M[x][y];
                    count++;
            
                }
                res[i][j] = sum/count;
            }
            
        }
        return res;
        
    }
};

File: find-duplicate-file-in-system/shaun-gy.java
/*

Runtime: 25 ms, faster than 75.33% of Java online submissions for Find Duplicate File in System.
Memory Usage: 49.7 MB, less than 100.00% of Java online submissions for Find Duplicate File in System.

shaun own idea with map<String,List<String>>
*/
class Solution {
    public List<List<String>> findDuplicate(String[] paths) {
        List<List<String>> res = new ArrayList<>();
        Map<String,List<String>> map = new HashMap<>();
        for (String path: paths) {
            String [] a = path.split(" ");
            String dir = a[0];
            for (int i = 1; i < a.length; i++) {
                int iLeft = a[i].indexOf("("), iRight = a[i].indexOf(")");
                
                String fn = a[i].substring(0,iLeft);
                String content = a[i].substring(iLeft+1,iRight);
                List<String> list = map.getOrDefault(content, new ArrayList<>());
                list.add(dir + "/" + fn);
                if (!map.containsKey(content)) map.put(content, list);
            }
        }
        
        for (List<String> list: map.values()) {  //valueSet()
            if (list.size() >= 2) res.add(list);
        }
        return res;
    }
}



//grandyang idea, shaun fully understands and rewrite own style
//runtime 124, faster than 56%, mem less than 21%
//cutting, string parse tokenizer, hash table, understand problem asks

